pnet_packet/src/dns.rs

```diff
@@ -467,6 +468,133 @@ pub struct DnsResponse {
     pub payload: Vec<u8>,
 }
 
+fn rname_length(packet: &DnsResponsePacket) -> usize {
+    let mut offset = 0;
+    let mut size = 0;
+    loop {
+        let label_len = packet.packet()[offset] as usize;
+        if label_len == 0 {
+            size += 1;
+            break;
+        }
+        if label_len & 0xC0 == 0xC0 {
+            size += 2;
+            break;
+        }
+        size += label_len + 1;
+        offset += label_len + 1;
+    }
+    size
+}
+
+// parse any string in the DNS message, using its unique encoding
+// Call this on DnsPacket, pass the string([u8]) to be decoded
+pub fn dns_parse_name(packet: &DnsPacket, coded_name: &Vec<u8>) -> String {
+    // First follow the path in the rname, except if it starts with a C0
+    // then move to using the offsets from the start
+    let start = packet.packet();
+    let mut name = coded_name.as_slice();
+    let mut rname = String::new();
+    let mut offset:usize = 0;
+
+    loop {
+        let label_len:u16 = name[offset] as u16;
+        if label_len == 0 {
+            break;
+        }
+        if (label_len & 0xC0) == 0xC0 {
+            let offset1 = ((label_len & 0x3F) as usize) << 8;
+            let offset2 = name[offset + 1] as usize;
+            offset =  offset1 + offset2;
+            // now change name
+            name = start;
+            continue;
+        }
+        if !rname.is_empty() {
+            rname.push('.');
+        }
+        rname.push_str(
+            str::from_utf8(&name[offset + 1..offset + 1 + label_len as usize])
+                .ok()
+                .unwrap(),
```

Comment:
I think this method should probably return a `Result` instead, rather than crashing on invalid utf8.

Reply:
Accepted. I overlooked this while copying the code from the qname funciton. I will make this change here and the corresponding test cases. Todo: we should use the punycode crate for the real validity check and return the true string. utf8 is not allowed in the DNS messages anyway.